vochain/scrutinizer/process.go

```diff
@@ -90,144 +142,127 @@ func (s *Scrutinizer) isLiveResultsProcess(processID []byte) (bool, error) {
 
-func (s *Scrutinizer) addLiveResultsProcess(pid []byte) {
-	log.Infof("add new process %x to live results", pid)
-	process, err := s.Storage.Get(s.Encode("liveProcess", pid))
-	if err != nil && err != badger.ErrKeyNotFound {
-		log.Error(err)
-		return
-	}
-	if len(process) > 0 {
-		return
+	// Add the entity to the indexer database
+	eid := p.GetEntityId()
+	if err := s.db.Upsert(eid, &Entity{ID: eid, CreationTime: currentBlockTime}); err != nil {
+		return err
 	}
-	// Create empty process and store it
-	if _, err := s.newEmptyLiveProcess(pid); err != nil {
-		log.Error(err)
-	}
-}
 
-// Encode encodes scrutinizer specific data adding a prefix for its inner database
-func (s *Scrutinizer) Encode(t string, data []byte) []byte {
-	switch t {
-	case "entity":
-		return append([]byte{types.ScrutinizerEntityPrefix}, data...)
-	case "liveProcess":
-		return append([]byte{types.ScrutinizerLiveProcessPrefix}, data...)
-	case "results":
-		return append([]byte{types.ScrutinizerResultsPrefix}, data...)
-	case "processEnding":
-		return append([]byte{types.ScrutinizerProcessEndingPrefix}, data...)
-	}
-	panic("scrutinizer encode type not known")
+	/*
+		// Make a copy of the protobuf types
+		envtype := proto.Clone(p.GetEnvelopeType()).(*models.EnvelopeType)
+		if envtype == nil {
+			envtype = new(models.EnvelopeType)
+		}
+		mode := proto.Clone(p.GetMode()).(*models.ProcessMode)
+		if mode == nil {
+			mode = new(models.ProcessMode)
+		}
+		voteopt := proto.Clone(p.GetVoteOptions()).(*models.ProcessVoteOptions)
+		if voteopt == nil {
+			voteopt = new(models.ProcessVoteOptions)
+		}
+	*/
+	// Create process in the indexer database
+	return s.db.Insert(pid, &Process{
+		ID:           pid,
+		EntityID:     eid,
+		StartBlock:   p.GetStartBlock(),
+		EndBlock:     p.GetBlockCount() + p.GetStartBlock(),
+		CensusRoot:   p.GetCensusRoot(),
+		CensusURI:    p.GetCensusURI(),
+		CensusOrigin: int32(p.GetCensusOrigin()),
+		Status:       int32(p.GetStatus()),
+		Namespace:    p.GetNamespace(),
+		PrivateKeys:  p.EncryptionPrivateKeys,
+		PublicKeys:   p.EncryptionPublicKeys,
+		Envelope:     p.GetEnvelopeType(),
+		Mode:         p.GetMode(),
+		VoteOpts:     p.GetVoteOptions(),
+		CreationTime: currentBlockTime,
+	})
 }
 
-func (s *Scrutinizer) addEntity(eid, pid []byte) {
-	// TODO(mvdan): use a prefixed database
-	storagekey := s.Encode("entity", eid)
-	processList, err := s.Storage.Get(storagekey)
-	if err != nil && err != badger.ErrKeyNotFound {
-		log.Errorf("addEntity: %s", err)
-		return
-	}
-	if err == badger.ErrKeyNotFound {
-		log.Infof("added new entity %x to scrutinizer", eid)
-	}
-	if len(pid) != types.ProcessIDsize {
-		log.Errorf("addEntity: pid size is not correct, got %d", len(pid))
-		return
+// updateProcess synchronize those fields that can be updated on a existing process
+// with the information obtained from the Vochain state
+func (s *Scrutinizer) updateProcess(pid []byte) error {
```

Comment:
This function is called upon some specific events happening on the Vochain state like onCancel() or OnRevealKeys().
The OnRevealKeys event changes the process keys fields but not other fields and with updateProcess some of the unchanged fields are asigned again.
Can this function behaves depending on the event that is triggered? e.g. if the only field changed is Status there is no need to asign again other fields. Does it make sense ? I see this function too generic.

Reply:
I was hoping that badgerHold Update() function would only change those non-zero value defined fields, but that's not the case. So instead of adding more complexity (like doing specific calls for each type of event) to the code and thus more probability of adding bugs, I decided to use a generic approach since the costs of writing 1 field or 5 fields of a struct, once the badger index has been found, loaded to memory and so on, its almost negligible.

Reply:
Oki, fine with that :) Thank you